File: 2022/day4/main.py
import os
import sys

class Solution:
    def __init__(self):
        print(self.part1())
        print(self.part2())

    def part1(self):
        c = 0
        with open(os.path.join(sys.path[0], "data.txt"), "r") as file:
            lines = file.read().split("\n")
            for line in lines:
                w, x = line.split(",")[0].split("-")
                y, z = line.split(",")[1].split("-")
                if (int(w) <= int(y) and int(x) >= int(z)) or (int(y) <= int(w) and int(x) <= int(z)): c+= 1
        
        return c

    def part2(self):
        c = 0
        with open(os.path.join(sys.path[0], "data.txt"), "r") as file:
            lines = file.read().split("\n")
            for line in lines:
                w, x = line.split(",")[0].split("-")
                y, z = line.split(",")[1].split("-")
                if int(x) >= int(y) and int(w) <= int(z): c+= 1
            
        return c
if __name__ == "__main__":
    Solution()

File: 2022/day4/main.rs
use std::fs::File;
use std::io::{prelude::*, BufReader};

fn part1() -> i32 {
    let file = File::open("data.txt").expect("Unable to open the file");
    let reader = BufReader::new(file);

    let mut count = 0;
    for line in reader.lines() {
        match line {
            Ok(val) => {
                let p: Vec<&str> = val.split(",").collect();
                let a: Vec<&str> = p[0].split("-").collect();
                let b: Vec<&str> = p[1].split("-").collect();
                if a[0].parse::<i32>().unwrap() <= b[0].parse::<i32>().unwrap() && a[1].parse::<i32>().unwrap() >= b[1].parse::<i32>().unwrap() 
                    || b[0].parse::<i32>().unwrap() <= a[0].parse::<i32>().unwrap() && a[1].parse::<i32>().unwrap() <= b[1].parse::<i32>().unwrap() {
                    count+=1;
                }

            },
            Err(_e) => {}
        };


    }
    return count;
}

fn part2() -> i32 {
    let file = File::open("data.txt").expect("Unable to open the file");
    let reader = BufReader::new(file);

    let mut count = 0;
    for line in reader.lines() {
        match line {
            Ok(val) => {
                let p: Vec<&str> = val.split(",").collect();
                let a: Vec<&str> = p[0].split("-").collect();
                let b: Vec<&str> = p[1].split("-").collect();
                if a[1].parse::<i32>().unwrap() >= b[0].parse::<i32>().unwrap() && a[0].parse::<i32>().unwrap() <= b[1].parse::<i32>().unwrap() {
                    count+=1;
                }

            },
            Err(_e) => {}
        };


    }
    return count;
}

fn main() {
    println!("{}", part1());
    println!("{}", part2());
}
